controller/src/lib.rs

```diff
@@ -2,10 +2,605 @@
+use tokio::time::Instant;
+use transceiver_messages::message;
+use transceiver_messages::message::Header;
+use transceiver_messages::message::HostRequest;
+use transceiver_messages::message::HostResponse;
+use transceiver_messages::message::Message;
+use transceiver_messages::message::MessageBody;
+use transceiver_messages::message::Status;
+use transceiver_messages::mgmt::MemoryRead;
+use transceiver_messages::Error as MessageError;
+use transceiver_messages::ModuleId;
+use transceiver_messages::ADDR;
+use transceiver_messages::MAX_PAYLOAD_SIZE;
+use transceiver_messages::PORT;
+
+/// An error related to managing the transceivers.
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("Error in transceiver control protocol: {0:?}")]
+    Protocol(#[from] transceiver_messages::Error),
+
+    #[error("Network or I/O error: {0}")]
+    Io(#[from] std::io::Error),
+
+    #[error("Message type requires data, but none provided")]
+    MessageRequiresData,
+
+    #[error("A serialization error occurred: {0}")]
+    SerDes(hubpack::Error),
+}
+
+// A request sent from host to SP, possibly with trailing data.
+#[derive(Clone, Debug)]
+struct HostRpcRequest {
+    pub message: Message,
+    pub data: Option<Vec<u8>>,
+}
+
+// A response sent from SP to host, possibly with trailing data.
+#[derive(Clone, Debug)]
+#[allow(dead_code)]
+struct SpRpcResponse {
+    pub message: Message,
+    pub data: Option<Vec<u8>>,
+}
+
+/// A request sent from SP to host, possibly with trailing data.
+#[derive(Clone, Debug)]
+pub struct SpRpcRequest {
+    pub message: Message,
+    pub data: Option<Vec<u8>>,
+}
+
+/// A response sent from host to SP, possibly with trailing data.
+#[derive(Clone, Debug)]
+pub struct HostRpcResponse {
+    pub message: Message,
+    pub data: Option<Vec<u8>>,
+}
+
+// A request from host to SP that has not yet been completed.
+#[derive(Debug)]
+struct OutstandingHostRequest {
+    // The actual request object we're sending. It's stored so that we can
+    // resend it if needed.
+    request: HostRpcRequest,
+    // The channel on which the eventual reply will be sent.
+    response_tx: oneshot::Sender<Result<SpRpcResponse, Error>>,
+    // The time we last sent the request, used during retries.
+    last_sent: Instant,
+}
+
+// We limit ourselves to a single outstanding request in either direction at
+// this point.
+const NUM_OUTSTANDING_REQUESTS: usize = 1;
 
 /// A type for controlling transceiver modules on a Sidecar.
 #[derive(Debug)]
-pub struct Controller;
+pub struct Controller {
+    _log: Logger,
+    message_id: AtomicU64,
+
+    // Channel onto which requests from the host to SP are sent.
+    //
+    // `io_task` owns the receiving end of this, and actually sends out the
+    // messages to the SP.
+    outgoing_request_tx: mpsc::Sender<OutstandingHostRequest>,
+
+    // The task handling actual network IO with the peer.
+    io_task: JoinHandle<()>,
+
+    // The task handling actual network IO with the peer.
```

Comment:
Same docstring comment as above?

Reply:
...Still seems the same as above?  I don't think `io_task` and `request_task` are doing exactly the same thing.

Reply:
D'oh, got lost in the local changes. Thanks, will fix now.

Reply:
Fixed in 95ab907